src/Compilers/CSharp/Test/Symbol/Symbols/MissingSpecialMember.cs

```diff
@@ -1529,7 +1496,7 @@ struct S
     private string str;
     public S(char chr) { this.str = chr.ToString(); }
     public S(string str) { this.str = str; }
-    public static S operator + (S x, S y) { return new S('(' + x.str + '+' + y.str + ')'); }
+    public static S operator + (S x, S y) { return new S(x.str + '+' + y.str); }
```

Comment:
This one's interesting. If we try and emit a call to `string.Concat(string, string, string)` but that member is missing, we'll emit `string.Missing(string, string, string)` instead (as an error type). However, we then go on to concat more things. We don't recognise the call to `string.Missing` as a valid call to an overload of `string.Concat`, and we don't know its return type, so we then go on to generate `string.Concat(string.Missing('(', x.str, '+')?.ToString(), y.str, ')')`. However, that second `string.Concat` is also missing, so we end up with `string.Missing(string.Missing('(', x.str, '+')?.ToString(), y.str, ')')` and two diagnostics.

This wasn't triggered previously. Since the return type of `string.Missing` isn't known, the second call to `string.Concat` we generated would be `string.Concat(object, object, object)` which wasn't missing. We didn't have a test which checked whether `string.Concat(object, object, object)` was missing by passing it exactly 5 parameters.

I tweaked the test to reduce the number of parameters, which avoids the issue. Do we want to fix it "properly"? If so, how?

Reply:
IMO not a big deal for scenarios that involve a malformed core lib to give a few spurious diagnostics, as long as they're not actively misleading.

Isn't the method used in practice actually `string.Concat(string[])`?

---
In reply to: [275783740](https://github.com/dotnet/roslyn/pull/35006#discussion_r275783740) [](ancestors = 275783740)

Reply:
:+1: 

We lower string concatenation by dealing with each binary `+` in turn. So if we have `A + B + C`, we first start with `A + B` and lower that to `string.Concat(A, B)`, so the whole expression is now `string.Concat(A, B) + C`. Then we visit the remaining `+`, see that the left-hand side is a call to `string.Concat`, extract the `A` and `B` from it, and turn the whole thing into `string.Concat(A, B, C)`.

So on the way to `string.Concat(string[5])`, we also emit calls to `string.Concat(string, string)`, `string.Concat(string, string, string)`, etc (when are then all removed when we look at the subsequent `+`).

The problem here is that we get as far as `string.Concat('(', x.str, '+') + y.str + ')'`, and because that method is missing, we emit something like `string.Missing('(', x.str, '+')`. When we come to lower `string.Missing('(', x.str, '+') + y.str`, we don't recognise that the left-hand side is a call to `string.Concat` that we can extract the arguments from, and instead deal with it in the same way as any other expression.